File: src/main/java/no/nav/veilarboppfolging/kafka/KvpPeriode.java
package no.nav.veilarboppfolging.kafka;


import lombok.*;
import no.nav.common.types.identer.AktorId;

import java.time.ZonedDateTime;

@Value
@NoArgsConstructor(force = true)
@AllArgsConstructor
@Builder
@With
public class KvpPeriode {
    KvpPeriodeEventType event;
    String aktorId;
    String enhetId;
    KvpStartet startet;
    KvpAvsluttet avsluttet;

    public static KvpPeriode start(AktorId aktorId, String enhetId, String opprettetAv, ZonedDateTime startDato, String begrunnelse) {
        var kvpStartet = new KvpStartet(opprettetAv, startDato, begrunnelse);
        return new KvpPeriode(KvpPeriodeEventType.STARTET, aktorId.get(), enhetId, kvpStartet, null);
    }

    public KvpPeriode avslutt(String avsluttetAv, ZonedDateTime avsluttetTidspunkt, String begrunnelse) {
        return this.withAvsluttet(new KvpAvsluttet(avsluttetAv, avsluttetTidspunkt, begrunnelse)).withEvent(KvpPeriodeEventType.AVSLUTTET);
    }
}

File: src/main/kotlin/no/nav/veilarboppfolging/kafka/KvpPeriode.kt
package no.nav.veilarboppfolging.kafka

import no.nav.common.types.identer.AktorId
import java.time.ZonedDateTime

data class KvpPeriode(
    val event: KvpPeriodeEventType,
    val aktorId: String,
    val enhetId: String,
    val startet: KvpStartet,
    val avsluttet: KvpAvsluttet?
) {

    fun tilAvsluttetKvpPeriode(avsluttetAv: String, avsluttetTidspunkt: ZonedDateTime, begrunnelse: String): KvpPeriode {
        return this.copy(
            avsluttet = KvpAvsluttet(avsluttetAv, avsluttetTidspunkt, begrunnelse),
            event = KvpPeriodeEventType.AVSLUTTET
        )
    }

    companion object {
        fun start(
            aktorId: AktorId,
            enhetId: String,
            opprettetAv: String,
            startDato: ZonedDateTime,
            begrunnelse: String?
        ): KvpPeriode {
            val kvpStartet = KvpStartet(opprettetAv, startDato, begrunnelse)
            return KvpPeriode(KvpPeriodeEventType.STARTET, aktorId.get(), enhetId, kvpStartet, null)
        }
    }
}


